File: deno-debugger/scripts/breadcrumbs.ts
/**
 * Investigation breadcrumbs tracking system.
 *
 * Records the investigation journey:
 * - Hypotheses formed
 * - Tests conducted
 * - Findings discovered
 * - Decisions made
 * - Data captured
 *
 * Enables:
 * - Timeline generation
 * - Reproducible investigations
 * - Comprehensive reporting
 */

import type { Breadcrumb, BreadcrumbType } from "./types.ts";

export interface BreadcrumbData {
  investigationName: string;
  startTime: string;
  breadcrumbs: Breadcrumb[];
  summary: InvestigationSummary;
}

export interface InvestigationSummary {
  investigationName: string;
  startTime: string;
  durationSeconds: number;
  breadcrumbCount: number;
  typeCounts: Record<string, number>;
}

export class Breadcrumbs {
  public investigationName: string;
  public breadcrumbs: Breadcrumb[] = [];
  public startTime: Date;

  constructor(investigationName?: string) {
    this.startTime = new Date();
    this.investigationName = investigationName ||
      `investigation_${this.startTime.toISOString().replace(/[:.]/g, "-")}`;
  }

  private addBreadcrumb(
    type: BreadcrumbType,
    description: string,
    data?: Record<string, unknown>,
    tags?: string[],
  ): Breadcrumb {
    const breadcrumb: Breadcrumb = {
      timestamp: new Date().toISOString(),
      type,
      description,
      data: data || {},
      tags: tags || [],
    };
    this.breadcrumbs.push(breadcrumb);
    return breadcrumb;
  }

  addHypothesis(description: string, rationale?: string, tags?: string[]): Breadcrumb {
    const details: Record<string, unknown> = {};
    if (rationale) {
      details.rationale = rationale;
    }
    return this.addBreadcrumb("hypothesis", description, details, tags);
  }

  addTest(
    testName: string,
    description: string,
    details?: Record<string, unknown>,
    tags?: string[],
  ): Breadcrumb {
    const testDetails = { ...details };
    testDetails.test_name = testName;
    return this.addBreadcrumb("test", description, testDetails, tags);
  }

  addFinding(
    finding: string,
    data?: Record<string, unknown>,
    severity?: "info" | "warning" | "critical",
    tags?: string[],
  ): Breadcrumb {
    const details = { ...data };
    if (severity) {
      details.severity = severity;
    }
    return this.addBreadcrumb("finding", finding, details, tags);
  }

  addDecision(
    decision: string,
    rationale: string,
    alternatives?: string[],
    tags?: string[],
  ): Breadcrumb {
    const details: Record<string, unknown> = { rationale };
    if (alternatives) {
      details.alternatives = alternatives;
    }
    return this.addBreadcrumb("decision", decision, details, tags);
  }

  addData(
    dataType: string,
    location: string,
    description?: string,
    metadata?: Record<string, unknown>,
    tags?: string[],
  ): Breadcrumb {
    const details = { ...metadata };
    details.data_type = dataType;
    details.location = location;

    const desc = description || `${dataType} captured at ${location}`;
    return this.addBreadcrumb("finding", desc, details, tags);
  }

  addNote(note: string, tags?: string[]): Breadcrumb {
    return this.addBreadcrumb("finding", note, {}, tags);
  }

  addQuestion(question: string, answer?: string, tags?: string[]): Breadcrumb {
    const details: Record<string, unknown> = {};
    if (answer) {
      details.answer = answer;
    }
    return this.addBreadcrumb("hypothesis", question, details, tags);
  }

  answerQuestion(questionIndex: number, answer: string): void {
    if (questionIndex >= 0 && questionIndex < this.breadcrumbs.length) {
      const crumb = this.breadcrumbs[questionIndex];
      if (crumb.type === "hypothesis") {
        crumb.data = crumb.data || {};
        crumb.data.answer = answer;
        crumb.data.answered_at = new Date().toISOString();
      }
    }
  }

  getTimeline(): Breadcrumb[] {
    return this.breadcrumbs;
  }

  getByType(type: BreadcrumbType): Breadcrumb[] {
    return this.breadcrumbs.filter((bc) => bc.type === type);
  }

  getByTag(tag: string): Breadcrumb[] {
    return this.breadcrumbs.filter((bc) => bc.tags?.includes(tag));
  }

  getSummary(): InvestigationSummary {
    const now = new Date();
    const duration = (now.getTime() - this.startTime.getTime()) / 1000;

    const types: BreadcrumbType[] = ["hypothesis", "test", "finding", "decision"];
    const typeCounts: Record<string, number> = {};
    for (const type of types) {
      typeCounts[type] = this.getByType(type).length;
    }

    return {
      investigationName: this.investigationName,
      startTime: this.startTime.toISOString(),
      durationSeconds: duration,
      breadcrumbCount: this.breadcrumbs.length,
      typeCounts,
    };
  }

  toOrgTimeline(): string {
    const lines: string[] = [
      `* Investigation Timeline: ${this.investigationName}`,
      `  :PROPERTIES:`,
      `  :START_TIME: ${this.startTime.toISOString()}`,
      `  :END:`,
      "",
    ];

    for (let i = 0; i < this.breadcrumbs.length; i++) {
      const bc = this.breadcrumbs[i];
      const ts = new Date(bc.timestamp);
      const orgTs = ts.toISOString().replace("T", " ").substring(0, 16);

      const icon: Record<string, string> = {
        "hypothesis": "❓",
        "test": "🧪",
        "finding": "🔍",
        "decision": "⚡",
      };

      lines.push(`** ${icon[bc.type] || "•"} ${bc.type.toUpperCase()}: ${bc.description}`);
      lines.push(`   [${orgTs}]`);

      if (bc.data && Object.keys(bc.data).length > 0) {
        lines.push("   :PROPERTIES:");
        for (const [key, value] of Object.entries(bc.data)) {
          const valStr = typeof value === "object" ? JSON.stringify(value, null, 2) : String(value);
          lines.push(`   :${key.toUpperCase()}: ${valStr}`);
        }
        lines.push("   :END:");
      }

      if (bc.tags && bc.tags.length > 0) {
        lines.push(`   Tags: ${bc.tags.join(", ")}`);
      }

      lines.push("");
    }

    return lines.join("\n");
  }

  toMarkdownTimeline(): string {
    const lines: string[] = [
      `# Investigation Timeline: ${this.investigationName}`,
      `Started: ${this.startTime.toISOString()}`,
      "",
    ];

    for (let i = 0; i < this.breadcrumbs.length; i++) {
      const bc = this.breadcrumbs[i];
      const ts = new Date(bc.timestamp);

      const icon: Record<string, string> = {
        "hypothesis": "❓",
        "test": "🧪",
        "finding": "🔍",
        "decision": "⚡",
      };

      lines.push(
        `## ${i + 1}. ${icon[bc.type] || "•"} ${bc.type.toUpperCase()}: ${bc.description}`,
      );
      lines.push(`*${ts.toISOString().replace("T", " ").substring(0, 19)}*`);
      lines.push("");

      if (bc.data && Object.keys(bc.data).length > 0) {
        for (const [key, value] of Object.entries(bc.data)) {
          lines.push(`- **${key}**: ${value}`);
        }
        lines.push("");
      }

      if (bc.tags && bc.tags.length > 0) {
        lines.push(`Tags: \`${bc.tags.join(", ")}\``);
        lines.push("");
      }
    }

    return lines.join("\n");
  }

  async save(filePath: string): Promise<void> {
    const data: BreadcrumbData = {
      investigationName: this.investigationName,
      startTime: this.startTime.toISOString(),
      breadcrumbs: this.breadcrumbs,
      summary: this.getSummary(),
    };

    await Deno.writeTextFile(filePath, JSON.stringify(data, null, 2));
    console.log(`Breadcrumbs saved to ${filePath}`);
  }

  static async load(filePath: string): Promise<Breadcrumbs> {
    const content = await Deno.readTextFile(filePath);
    const data = JSON.parse(content) as BreadcrumbData;

    const bc = new Breadcrumbs(data.investigationName);
    bc.startTime = new Date(data.startTime);
    bc.breadcrumbs = data.breadcrumbs;

    return bc;
  }
}

// ============================================================================
// CLI Usage
// ============================================================================

if (import.meta.main) {
  console.log("Breadcrumbs - Example Usage");
  console.log("============================");
  console.log();
  console.log("// Initialize tracking");
  console.log("const bc = new Breadcrumbs('memory_leak_investigation');");
  console.log();
  console.log("// Record hypothesis");
  console.log("bc.addHypothesis(");
  console.log("  'Upload handler retains file buffers in closure',");
  console.log("  'Heap shows growing ArrayBuffer objects'");
  console.log(");");
  console.log();
  console.log("// Record test");
  console.log("bc.addTest(");
  console.log("  'heap_snapshot_comparison',");
  console.log("  'Compare heap before and after 10 uploads',");
  console.log("  { snapshots: ['before.heap', 'after.heap'] }");
  console.log(");");
  console.log();
  console.log("// Record finding");
  console.log("bc.addFinding(");
  console.log("  'ArrayBuffer objects growing by 50MB per upload',");
  console.log("  { growth_rate: '50MB/upload', object_count: 1250 },");
  console.log("  'critical'");
  console.log(");");
  console.log();
  console.log("// Save and export");
  console.log("await bc.save('investigation.json');");
  console.log("const timeline = bc.toMarkdownTimeline();");
  console.log("console.log(timeline);");
}

File: skill/scripts/breadcrumbs.py
"""
Investigation breadcrumbs tracking system.

Records the investigation journey:
- Hypotheses formed
- Tests conducted
- Findings discovered
- Decisions made
- Data captured

Enables:
- Timeline generation
- Reproducible investigations
- Comprehensive reporting
"""

import json
from datetime import datetime
from typing import Any, Dict, List, Optional
from dataclasses import dataclass, asdict
from enum import Enum


class BreadcrumbType(Enum):
    """Types of breadcrumbs that can be recorded."""
    HYPOTHESIS = "hypothesis"
    TEST = "test"
    FINDING = "finding"
    DECISION = "decision"
    DATA = "data"
    NOTE = "note"
    QUESTION = "question"


@dataclass
class Breadcrumb:
    """A single breadcrumb in the investigation."""
    timestamp: str
    type: str
    description: str
    details: Dict[str, Any]
    tags: List[str]

    def to_dict(self) -> Dict:
        """Convert to dictionary."""
        return asdict(self)


class Breadcrumbs:
    """
    Investigation breadcrumbs tracker.

    Usage:
        bc = Breadcrumbs()
        bc.add_hypothesis("Upload handler may retain file buffers")
        bc.add_test("heap_snapshot_comparison", {"files": ["before.heap", "after.heap"]})
        bc.add_finding("ArrayBuffer growing by 50MB per upload", {"growth_rate": "50MB"})
        bc.save("investigation.json")
    """

    def __init__(self, investigation_name: Optional[str] = None):
        self.investigation_name = investigation_name or f"investigation_{datetime.now().strftime('%Y%m%d_%H%M%S')}"
        self.breadcrumbs: List[Breadcrumb] = []
        self.start_time = datetime.now()

    def _add_breadcrumb(
        self,
        crumb_type: BreadcrumbType,
        description: str,
        details: Optional[Dict[str, Any]] = None,
        tags: Optional[List[str]] = None
    ):
        """Internal method to add a breadcrumb."""
        breadcrumb = Breadcrumb(
            timestamp=datetime.now().isoformat(),
            type=crumb_type.value,
            description=description,
            details=details or {},
            tags=tags or []
        )
        self.breadcrumbs.append(breadcrumb)
        return breadcrumb

    def add_hypothesis(self, description: str, rationale: Optional[str] = None, tags: Optional[List[str]] = None):
        """
        Record a hypothesis about what might be wrong.

        Args:
            description: The hypothesis statement
            rationale: Why you think this might be the issue
            tags: Optional tags for categorization
        """
        details = {}
        if rationale:
            details['rationale'] = rationale

        return self._add_breadcrumb(BreadcrumbType.HYPOTHESIS, description, details, tags)

    def add_test(
        self,
        test_name: str,
        description: str,
        details: Optional[Dict[str, Any]] = None,
        tags: Optional[List[str]] = None
    ):
        """
        Record a test being conducted.

        Args:
            test_name: Name of the test
            description: What the test does
            details: Additional details (files, parameters, etc.)
            tags: Optional tags
        """
        test_details = details or {}
        test_details['test_name'] = test_name

        return self._add_breadcrumb(BreadcrumbType.TEST, description, test_details, tags)

    def add_finding(
        self,
        finding: str,
        data: Optional[Dict[str, Any]] = None,
        severity: Optional[str] = None,
        tags: Optional[List[str]] = None
    ):
        """
        Record a finding or discovery.

        Args:
            finding: What was discovered
            data: Supporting data
            severity: Optional severity (low, medium, high, critical)
            tags: Optional tags
        """
        details = data or {}
        if severity:
            details['severity'] = severity

        return self._add_breadcrumb(BreadcrumbType.FINDING, finding, details, tags)

    def add_decision(
        self,
        decision: str,
        rationale: str,
        alternatives: Optional[List[str]] = None,
        tags: Optional[List[str]] = None
    ):
        """
        Record a decision made during investigation.

        Args:
            decision: What decision was made
            rationale: Why this decision was made
            alternatives: Other options considered
            tags: Optional tags
        """
        details = {
            'rationale': rationale
        }
        if alternatives:
            details['alternatives'] = alternatives

        return self._add_breadcrumb(BreadcrumbType.DECISION, decision, details, tags)

    def add_data(
        self,
        data_type: str,
        location: str,
        description: Optional[str] = None,
        metadata: Optional[Dict[str, Any]] = None,
        tags: Optional[List[str]] = None
    ):
        """
        Record data captured during investigation.

        Args:
            data_type: Type of data (snapshot, profile, log, etc.)
            location: Where the data is stored
            description: Optional description
            metadata: Additional metadata
        """
        details = metadata or {}
        details['data_type'] = data_type
        details['location'] = location

        desc = description or f"{data_type} captured at {location}"

        return self._add_breadcrumb(BreadcrumbType.DATA, desc, details, tags)

    def add_note(self, note: str, tags: Optional[List[str]] = None):
        """Record a general note or observation."""
        return self._add_breadcrumb(BreadcrumbType.NOTE, note, {}, tags)

    def add_question(
        self,
        question: str,
        answer: Optional[str] = None,
        tags: Optional[List[str]] = None
    ):
        """
        Record a question (and optionally its answer).

        Args:
            question: The question
            answer: Optional answer (can be added later)
            tags: Optional tags
        """
        details = {}
        if answer:
            details['answer'] = answer

        return self._add_breadcrumb(BreadcrumbType.QUESTION, question, details, tags)

    def answer_question(self, question_index: int, answer: str):
        """Add an answer to a previously asked question."""
        if 0 <= question_index < len(self.breadcrumbs):
            crumb = self.breadcrumbs[question_index]
            if crumb.type == BreadcrumbType.QUESTION.value:
                crumb.details['answer'] = answer
                crumb.details['answered_at'] = datetime.now().isoformat()

    def get_timeline(self) -> List[Breadcrumb]:
        """Get all breadcrumbs in chronological order."""
        return self.breadcrumbs

    def get_by_type(self, crumb_type: BreadcrumbType) -> List[Breadcrumb]:
        """Get all breadcrumbs of a specific type."""
        return [bc for bc in self.breadcrumbs if bc.type == crumb_type.value]

    def get_by_tag(self, tag: str) -> List[Breadcrumb]:
        """Get all breadcrumbs with a specific tag."""
        return [bc for bc in self.breadcrumbs if tag in bc.tags]

    def get_summary(self) -> Dict[str, Any]:
        """Get a summary of the investigation."""
        duration = datetime.now() - self.start_time

        type_counts = {}
        for bc_type in BreadcrumbType:
            type_counts[bc_type.value] = len(self.get_by_type(bc_type))

        return {
            'investigation_name': self.investigation_name,
            'start_time': self.start_time.isoformat(),
            'duration_seconds': duration.total_seconds(),
            'breadcrumb_count': len(self.breadcrumbs),
            'type_counts': type_counts,
        }

    def to_org_timeline(self) -> str:
        """
        Generate an Org mode timeline of the investigation.

        Returns:
            Org mode formatted timeline
        """
        lines = [
            f"* Investigation Timeline: {self.investigation_name}",
            f"  :PROPERTIES:",
            f"  :START_TIME: {self.start_time.isoformat()}",
            f"  :END:",
            ""
        ]

        for i, bc in enumerate(self.breadcrumbs, 1):
            # Parse timestamp for org format
            ts = datetime.fromisoformat(bc.timestamp)
            org_ts = ts.strftime("[%Y-%m-%d %a %H:%M]")

            # Create heading
            icon = {
                'hypothesis': '❓',
                'test': '🧪',
                'finding': '🔍',
                'decision': '⚡',
                'data': '💾',
                'note': '📝',
                'question': '❔'
            }.get(bc.type, '•')

            lines.append(f"** {icon} {bc.type.upper()}: {bc.description}")
            lines.append(f"   {org_ts}")

            # Add details
            if bc.details:
                lines.append("   :PROPERTIES:")
                for key, value in bc.details.items():
                    # Format value for org properties
                    if isinstance(value, (list, dict)):
                        value = json.dumps(value, indent=2)
                    lines.append(f"   :{key.upper()}: {value}")
                lines.append("   :END:")

            # Add tags
            if bc.tags:
                lines.append(f"   Tags: {', '.join(bc.tags)}")

            lines.append("")

        return '\n'.join(lines)

    def to_markdown_timeline(self) -> str:
        """
        Generate a Markdown timeline of the investigation.

        Returns:
            Markdown formatted timeline
        """
        lines = [
            f"# Investigation Timeline: {self.investigation_name}",
            f"Started: {self.start_time.isoformat()}",
            ""
        ]

        for i, bc in enumerate(self.breadcrumbs, 1):
            ts = datetime.fromisoformat(bc.timestamp)

            icon = {
                'hypothesis': '❓',
                'test': '🧪',
                'finding': '🔍',
                'decision': '⚡',
                'data': '💾',
                'note': '📝',
                'question': '❔'
            }.get(bc.type, '•')

            lines.append(f"## {i}. {icon} {bc.type.upper()}: {bc.description}")
            lines.append(f"*{ts.strftime('%Y-%m-%d %H:%M:%S')}*")
            lines.append("")

            # Add details
            if bc.details:
                for key, value in bc.details.items():
                    lines.append(f"- **{key}**: {value}")
                lines.append("")

            # Add tags
            if bc.tags:
                lines.append(f"Tags: `{', '.join(bc.tags)}`")
                lines.append("")

        return '\n'.join(lines)

    def save(self, file_path: str):
        """
        Save breadcrumbs to JSON file.

        Args:
            file_path: Output file path
        """
        data = {
            'investigation_name': self.investigation_name,
            'start_time': self.start_time.isoformat(),
            'breadcrumbs': [bc.to_dict() for bc in self.breadcrumbs],
            'summary': self.get_summary()
        }

        with open(file_path, 'w') as f:
            json.dump(data, f, indent=2)

        print(f"Breadcrumbs saved to {file_path}")

    @classmethod
    def load(cls, file_path: str) -> 'Breadcrumbs':
        """
        Load breadcrumbs from JSON file.

        Args:
            file_path: Input file path

        Returns:
            Breadcrumbs instance
        """
        with open(file_path, 'r') as f:
            data = json.load(f)

        bc = cls(investigation_name=data['investigation_name'])
        bc.start_time = datetime.fromisoformat(data['start_time'])

        for crumb_data in data['breadcrumbs']:
            bc.breadcrumbs.append(Breadcrumb(**crumb_data))

        return bc


if __name__ == '__main__':
    print("Breadcrumbs - Example Usage")
    print("============================")
    print()
    print("# Initialize tracking")
    print("bc = Breadcrumbs('memory_leak_investigation')")
    print()
    print("# Record hypothesis")
    print("bc.add_hypothesis(")
    print("    'Upload handler retains file buffers in closure',")
    print("    rationale='Heap shows growing ArrayBuffer objects'")
    print(")")
    print()
    print("# Record test")
    print("bc.add_test(")
    print("    'heap_snapshot_comparison',")
    print("    'Compare heap before and after 10 uploads',")
    print("    details={'snapshots': ['before.heap', 'after.heap']}")
    print(")")
    print()
    print("# Record finding")
    print("bc.add_finding(")
    print("    'ArrayBuffer objects growing by 50MB per upload',")
    print("    data={'growth_rate': '50MB/upload', 'object_count': 1250},")
    print("    severity='high'")
    print(")")
    print()
    print("# Save and export")
    print("bc.save('investigation.json')")
    print("timeline = bc.to_org_timeline()")
    print("print(timeline)")
